include/deal.II/base/work_stream.h

```diff
@@ -663,6 +663,163 @@ namespace WorkStream
 #  endif // DEAL_II_WITH_TBB
 
 
+
+#  ifdef DEAL_II_WITH_TASKFLOW
+    /**
+     * Mostly a copy of the 2nd implementation of the Workstream paper taking
+     * advantage of thread local lists for re-use. Uses taskflow for task
+     * scheduling rather than TBB. Currently does not support chunking.
+     */
+
+    namespace taskflow_no_coloring
+    {
+      template <typename Worker,
+                typename Copier,
+                typename Iterator,
+                typename ScratchData,
+                typename CopyData>
+
+      /**
+       *  The last two arguments in this function are for chunking support which
+       * currently does not exist but ideally will later. For now they are
+       * ignored but still here to permit existing programs to function
+       */
+      void
+      run(const Iterator                          &begin,
+          const typename identity<Iterator>::type &end,
+          Worker                                   worker,
+          Copier                                   copier,
+          const ScratchData                       &sample_scratch_data,
+          const CopyData                          &sample_copy_data,
+          const unsigned int queue_length = 2 * MultithreadInfo::n_threads(),
+          const unsigned int chunk_size   = 8)
+
+      {
+        struct ScratchDataObjects
+        {
+          std::unique_ptr<ScratchData> scratch_data;
+          bool                         currently_in_use;
+
+          /**
+           * Default constructor.
+           */
+          ScratchDataObjects()
+            : currently_in_use(false)
+          {}
+
+          ScratchDataObjects(std::unique_ptr<ScratchData> &&p,
+                             const bool                     in_use)
+            : scratch_data(std::move(p))
+            , currently_in_use(in_use)
+          {}
+
+          // Provide a copy constructor that actually doesn't copy the
+          // internal state. This makes handling ScratchAndCopyDataObjects
+          // easier to handle with STL containers.
+          ScratchDataObjects(const ScratchDataObjects &)
+            : currently_in_use(false)
+          {}
+        };
```

Comment:
There is a class in the TBB implementation called `ScratchDataObject` (singular) that is very similar except (i) the singular vs plural in the name, (ii) the other class has a defaulted move constructor. Why does yours use the plural form rather than the singular, and why does it not have a move constructor?

Reply:
Separately, if they are *intended* to be equivalent, could we avoid the code duplication by moving them into a place that is accessible by both implementations?